File: 3.longest-substring-without-repeating-characters.java
/*
 * @lc app=leetcode id=3 lang=java
 *
 * [3] Longest Substring Without Repeating Characters
 */
// @lc code=start
import java.util.Map;
import java.util.HashMap;
class Solution {
    public int lengthOfLongestSubstring(String s) {
        // Map<Character,Integer> map = new HashMap<Character,Integer>(256);
        int[] map = new int[256];
        Arrays.fill(map,-1);
        int l = 0, r = 0, mLen = 0;
        for(r = 0; r<s.length();r++){
            char ch = s.charAt(r);
            // int mVal = map.getOrDefault(ch,-1);
            int mVal = map[ch];
            if(mVal>=l){
                l = mVal+1;
            }
            mLen = Math.max(mLen,r-l+1);
            // map.put(ch,r);
            map[ch]=r;
        }
        return mLen;
    }
}
// @lc code=end

File: 3.longest-substring-without-repeating-characters.cpp
/*
 * @lc app=leetcode id=3 lang=cpp
 *
 * [3] Longest Substring Without Repeating Characters
 */
#include <array>
#include <string>
#include <map>
#include <set>
#include <unordered_map>
#include <unordered_set>
#include <vector>
using std::string;
using std::map;
using std::set;
using std::unordered_map;
using std::unordered_set;
using std::vector;
// @lc code=start
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int i=0,j=0,mLen=0;
        // unordered_map<char,int> map;
        std::array<int,sizeof(char)<<8> map;
        map.fill(-1);
        for(i=0;i<s.length();i++){
            char ch = s.at(i);
            // if(map.count(ch) && map[ch]>=j) j=map[ch]+1;
            if(map[ch]>=j) j=map[ch]+1;
            mLen = std::max(mLen, i-j+1);
            map[ch] = i;
        }
        return mLen;
    }
};
// @lc code=end

